controllers/pipelines/provider_controller.go

```diff
@@ -0,0 +1,220 @@
+
+	"k8s.io/apimachinery/pkg/runtime"
+
+	config "github.com/sky-uk/kfp-operator/apis/config/v1alpha6"
+	"github.com/sky-uk/kfp-operator/apis/pipelines"
+	pipelinesv1 "github.com/sky-uk/kfp-operator/apis/pipelines/v1alpha6"
+	"github.com/sky-uk/kfp-operator/controllers/pipelines/internal/logkeys"
+	appsv1 "k8s.io/api/apps/v1"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+const (
+	OwnerNameLabel         = "owner-name"
+	AppLabel               = "app"
+	ResourceHashAnnotation = "resource-hash"
+)
+
+type ProviderReconciler struct {
+	ResourceReconciler[*pipelinesv1.Provider]
+	Scheme *runtime.Scheme
+}
+
+func NewProviderReconciler(ec K8sExecutionContext, config config.KfpControllerConfigSpec) *ProviderReconciler {
+	return &ProviderReconciler{
+		ResourceReconciler: ResourceReconciler[*pipelinesv1.Provider]{
+			EC:     ec,
+			Config: config,
+		},
+		Scheme: ec.Scheme,
+	}
+}
+
+//+kubebuilder:rbac:groups=pipelines.kubeflow.org,resources=providers,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=pipelines.kubeflow.org,resources=providers/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=pipelines.kubeflow.org,resources=providers/finalizers,verbs=update
+//+kubebuilder:rbac:groups="",resources=events,verbs=create;patch
+
+func (r *ProviderReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	logger := log.FromContext(ctx)
+	startTime := time.Now()
+	logger.Info("reconciliation started", "request", req)
+
+	var provider = &pipelinesv1.Provider{}
+	if err := r.EC.Client.NonCached.Get(ctx, req.NamespacedName, provider); err != nil {
+		logger.Error(err, "unable to get provider")
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	logger.Info("found provider", "resource", provider)
+
+	desiredDeployment, err := constructDeployment(provider, *r.Config.DeepCopy())
+	if err != nil {
+		logger.Error(err, "unable to construct provider service deployment")
+		return ctrl.Result{}, err
+	}
+	if err := ctrl.SetControllerReference(provider, desiredDeployment, r.Scheme); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	logger.Info("constructed desired provider deployment", "deployment", desiredDeployment)
+
+	existingDeployment, err := r.getDeployment(ctx, *provider)
+	if err != nil && !apierrors.IsNotFound(err) {
+		logger.Error(err, "unable to get existing deployment")
+		return ctrl.Result{}, err
+	}
+
+	if existingDeployment != nil {
+		logger.Info("found existing provider deployment", "deployment", existingDeployment)
+
+		if err := setResourceHashAnnotation(existingDeployment); err != nil {
+			logger.Error(err, "unable to set resource hash annotation on existing deployment")
+			return ctrl.Result{}, err
+		}
+		if deploymentIsOutOfSync(existingDeployment, desiredDeployment) {
```

Comment:
```suggestion
		
		if deploymentIsOutOfSync(existingDeployment, desiredDeployment) {
```

Reply:
Done in https://github.com/sky-uk/kfp-operator/commit/8b421e577d6992d542f07803bb55b3e9f1cea44a